### ISSUE TITLE
De-thrift brave-core

### ISSUE THREAD

**Issue by adriancole** [openzipkin/brave#128]
generally decided at the tracing workshop that thrift has moved from a constant to an option.

Since brave was written when there was only one representation (thrift), it hosted the thrift objects in core. By moving span collectors to their own modules, we have focused the reporting function (including thrift encoding). Now, we have an opportunity to lighten the dependencies of brave to.. none?

The approach we have is to remove thrift encoding functions and supertypes from the `com.twitter.zipkin.gen` classes. At that point, these become plain java objects. We then make sure codepaths operate without a thrift dependency.

The reporting code paths will break because `com.twitter.zipkin.gen` classes are no longer thrift types. We need to introduce a class that converts these into zipkin-java types for the purpose of sending on the wire. This could make an encoding function which translates the spans to a supported representations.

```java
interface SpanEncoder {
  interface Factory {
    /** Returns an encoder for the given media type (ex. "application/json") or null if not found. */
    @Nullable
    SpanEncoder get(String mediaType);
  }

  byte[]/ByteBuffer/etc encode(Span in);
}
```

**Comment by CodingFabian** [comment]
+1

**Comment by CodingFabian** [comment]
I have worked on this internally, and now have a thrift-less version of brave-core
However there a couple of non-compatible changes:

* Thrift uses a dirty check for optional primitive field. Do we want this dirty check as well or are we going with Nullable boxed fields?

* Brave uses a binary annotation of type "bool" once: https://github.com/openzipkin/brave/blob/master/brave-core/src/main/java/com/github/kristofa/brave/AnnotationSubmitter.java#L113 Is this actually supported? I tend to change it to a string with contents "1". I would also make the BinaryAnnotation take and use a String rather than requiring the transformation to bytes.

* Model classes are java.io.Serializable, Cloneable, Comparable<Annotation>  - I would only retain the serializable attribute. (clone is broken, and the compare logic is strange somehow)

**Comment by adriancole** [comment]
Thanks for the progress!

>
>    -
>
>    Thrift uses a dirty check for optional primitive field. Do we want
>    this dirty check as well or are we going with Nullable boxed fields?
>
> The latter.. I think :)

>
>    -
>
>    Brave uses a binary annotation of type "bool" once:
>    https://github.com/openzipkin/brave/blob/master/brave-core/src/main/java/com/github/kristofa/brave/AnnotationSubmitter.java#L113
>    Is this actually supported? I tend to change it to a string with contents
>    "1". I would also make the BinaryAnnotation take and use a String rather
>    than requiring the transformation to bytes.
>
> Yes, this is a special case for recording the client or server addresses.
definitely in use
https://github.com/openzipkin/zipkin/blob/master/zipkin-thrift/src/main/thrift/com/twitter/zipkin/zipkinCore.thrift#L207


>
>    -
>
>    Model classes are java.io.Serializable, Cloneable, Comparable - I
>    would only retain the serializable attribute. (clone is broken, and the
>    compare logic is strange somehow)
>
> +1

**Comment by adriancole** [comment]
closed via #137
